### PR TITLE
Hero

### PR DESCRIPTION
**Pull request by helenvholmes**
No ticket for this work, part 1 of 2.

## **This PR does the following:**
- Abstracts together hero types 1&2 from [this Figma frame](https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=10968%3A5314). 
- This PR does not add types 3&4, and the tests are still insufficient... I've never written tests before so I might need to 1:1 with someone to get those up to snuff.
- The `SectionName` decorator also needs some work but felt out of scope for this ticket.

### Front End Review:
- [ ] View [Hero types in Storybook](http://localhost:9001/?path=/story/hero--hero-primary)
- [ ] Check against the Metronome documentation:
  - [ ] [Exhibition Header](http://themetronome.co/components/oh-64-exhibition-header/?fresh=true)
  - [ ] [Header with Image Right](http://themetronome.co/components/oh-30-header-with-image-right/?fresh=true)

### CODE REVIEW

**Comment by EdwinGuzman** [comment]
There are some linting issues which are causing the tests to fail.

File: src/react-components/src/components/00-base/Placeholder.tsx
Comment (hokei) [inline src/react-components/src/components/00-base/Placeholder.tsx:14]: I know this is really "do as I say and not as I do" but it would be really nice to have comments for all the components saying what they do.

File: src/react-components/src/components/00-base/SectionName.tsx
Comment (hokei) [inline src/react-components/src/components/00-base/SectionName.tsx:4]: I wonder if "research" is a thing that should be hardcoded, or if this is a thing that should be passed as a modifier from the consumer (so a Research product would pass "research" but a Locations product could pass "locations", etc)

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: ID generation is gonna get punted, the Filters ticket is large enough as is.

File: src/react-components/src/components/00-base/Placeholder.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/00-base/Placeholder.tsx:15]: These props aren't being used in this component.

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx:22]: Is it required for the `Heading` to have a blockName of `"hero"`? If that's the case, is it written down somewhere? Also, what happens if you forget to include this prop in `Heading`?

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: For short comments, I think you can write it in a single line and it'll work the same `/** Required heading element. */`.

File: src/react-components/src/components/03-organisms/Hero/HeroTypes.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/03-organisms/Hero/HeroTypes.tsx:1]: Very nice.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: Can you do `heroType === HeroType.Primary` instead of the string check? It'll be better to reduce strings in the code and use an existing variable.

File: src/react-components/src/components/00-base/SectionName.tsx
Comment (helenvholmes) [reply src/react-components/src/components/00-base/SectionName.tsx:4]: That's what I'd like to do but it felt out of scope for this ticket—completely agree though!

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx:22]: It is required, and it would fail as a result. How would something like this typically be handled?

File: src/react-components/src/components/00-base/Placeholder.tsx
Comment (helenvholmes) [reply src/react-components/src/components/00-base/Placeholder.tsx:14]: Updated!

File: src/react-components/src/components/00-base/SectionName.tsx
Comment (helenvholmes) [reply src/react-components/src/components/00-base/SectionName.tsx:4]: (To explain _why_ such a small thing wasn't done here, though, is that I'm not entirely sure what all of the site sections are yet. I could create another `enum` just to get this started and clean this up and add on once I've tracked that info down... I will do that, actually)

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: Updated the variable description!

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: Updated!

File: src/react-components/src/__tests__/components/Hero-test.tsx
Comment (hokei) [inline src/react-components/src/__tests__/components/Hero-test.tsx:14]: This test is good, but I'd have many more tests.  eg:  What happens if there's a HeroTypes.Primary but the user forgets to pass a background image? (which is optional) 
(that kind of stuff - what combinations of things could a user pass, and what kind of errors would be shown if the user passes in unexpected things?

You don't need to check anything that's checked by PropTypes or Typescript, but anything where it's, "On a Primary hero, an Image is required.  On a Secondary hero, a background image is ignored if passed, etc" I'd write a test for that.

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (hokei) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx:22]: if all heros should have blockname "hero", I would figure out how to incorporate it into the component itself rather than expecting users to to remember to pass it.

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx:17]: (may be out of scope) I'd love to figure out a way to make it clear that users are supposed to be only using this enum, and how to do that ... Does Typescript throw a useful error when passing in a string or does it let it through?

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx:4]: Stylistic, but I like to have these small, tightly coupled type classes in this component file rather than a separate one

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: change this to JSX.Element

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx:24]: What happens when you pass in a non-hex color?  Can we throw a useful error?

Comment (hokei) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx:24]: What happens when you pass in a color and you're not using an Exhibition hero?

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: If it's just text, I would use fragment:  
`<>{subtitleText}</>` rather than `dangerouslySetInnerHtml`
In this case since I think it's best that subtitle is a `JSX.Element` you can go {subtitle}

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: This is my bad originally, I thought strings were good enough to do this with, but if you, say, ever have a subtitle with a link in it, or a <b></b> in it, it wouldn't be possible, and it doesn't make sense to go to a PM and say "I can't make part of the subheader italicized"

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: is it possible to have a PRIMARY hero type with a "secondary" modifier?  That sounds ... confusing. If not, can we use the HeroType to decide the modifier?

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: nit: useful/cheap to say which one you got/is missing, could be helpful.

File: src/styles/01-atoms/links/link/_link.scss
Comment (hokei) [inline src/styles/01-atoms/links/link/_link.scss:2]: is there guidance on how to override this?

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: does this need to be printed every time Hero is used?

**Review by hokei — CHANGES_REQUESTED** [review]
happy to do a walkthrough on tests to better clarify!

File: src/styles/01-atoms/links/link/_link.scss
Comment (helenvholmes) [reply src/styles/01-atoms/links/link/_link.scss:2]: Not currently. This code definitely could use some work but I needed the latest and greatest from Reno for some of the Hero styling. Where do you think that guidance would live? Perhaps in a comment blob at the top of this file?

Comment (hokei) [reply src/styles/01-atoms/links/link/_link.scss:2]: some "Overriding Styles" document in Design System, I think.  The idea is that each product should have its style override sheets

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx:4]: If there are more times, it might be helpful to have a util "Types.tsx" file to hold all the types. Just another option.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: That's true. On the larger scale of this, it should be a written down requirement that this _can_ be italicized. Otherwise, you'd be coding for possible scenarios that might not be implemented.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: `dangerouslySetInnerHTML` can also be thought of as elements that will get rendered but you don't know where they are coming from. For example, maybe an admin added HTML tags to a description they want to render and that info is coming from an API.

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (EdwinGuzman) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx:17]: Passing in a string should throw an error since the `heroType` prop is defined as `HeroTypes`. I don't know of a clear way to let users know that it's supposed to be an enum, but maybe reading the component itself is enough?

File: src/styles/01-atoms/links/link/_link.scss
Comment (helenvholmes) [reply src/styles/01-atoms/links/link/_link.scss:2]: I'll create a separate ticket and wiki page for that work, then. It'll be blocked on upcoming color work/`SectionName.tsx` work.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: @EdwinGuzman, in Reno where this Exhibition header gets used, that's exactly the use case! Sounds like we have a JSX.Element that Crystal's created that would be better to use here, though, so I will switch it.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx:4]: I don't mind it being in a separate file so long as it's grouped within a folder. (Plus I think it keeps `Hero.tsx` from getting too long.) I'm a little more wary of a `Types.tsx` file simply because over time I think it can become a lot of code that doesn't feel as logically grouped as it would with its component.

I'm happy to move the `enum` into `Hero` if you feel strongly about it, Crystal—just want to be sure we're doing it the same way everywhere 😄

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: Good point. I wrote this before implementing the enum and didn't revisit it. This makes sense as cleanup.

File: src/styles/01-atoms/links/link/_link.scss
Comment (helenvholmes) [reply src/styles/01-atoms/links/link/_link.scss:2]: I've opened up #107 to capture log this work. Logistically I'm thinking it will happen after any `Template` work upcoming for RN.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (hokei) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx:4]: I don't feel strongly.  Agreed that we should do it the same way everywhere.  I think button types will need to be factored out soon so we can do it the same way.

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (hokei) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx:17]: OK!  In that case all we need is to be exporting `HeroType` in `index.ts`.

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx:22]: The idea is that all Headings should have the blockname hero, not all hero components (components have a base class of hero). This would happen anytime we pull in a JSX.Element into another component in a story like this. Is this solved anywhere else @hokei that you know of? I'm operating off my understanding within Reno ([example](https://github.com/NYPL/nypl-d8/blob/development/web/themes/custom/nypl_emulsify/components/_patterns/02-molecules/basic-card/basic-card.twig#L14)) but it could be that when including JSX elements there's a better way to do this.

Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx:22]: Talked about this with @hokei over Slack. Handling this particular problem (how sub-JSX.Elements get the correct blockNames within larger components) is out of scope for this ticket, so I'm going to leave this as-is for now.

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx:17]: Exported `HeroTypes` in `index.ts`.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx:4]: Leaving it as-is.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: I've updated this to use `<BodyText>`.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: Done

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: Removed

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx]: Done

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx:24]: While I think it would be nice to do some type-checking on colors (e.g., you pass in `#fffffff` by mistake, we could throw an error) I would rather not error on non-hex responses. `#fff`, `#ffffff`, and `white` are all valid CSS colors.

File: src/react-components/src/components/00-base/SectionName.tsx
Comment (hokei) [inline src/react-components/src/components/00-base/SectionName.tsx:4]: This seems pretty specific - does it need to be its own component?  If it's supposed to only be used by the ResearchNow hero, can we wrap the ResearchNow Hero Story with  
`<div className={bem('main', ['research'])} ></div> `?

Comment (helenvholmes) [reply src/react-components/src/components/00-base/SectionName.tsx:4]: It will be generic and used for all of the SectionNames stipulated here: https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=674%3A274

See former thread discussing why that work isn't being done yet: https://github.com/NYPL/nypl-design-system/pull/106/files#r401805076

File: src/styles/03-organisms/headers/_hero.scss
Comment (hokei) [inline src/styles/03-organisms/headers/_hero.scss]: where is this being used?

File: src/react-components/src/__tests__/components/Hero-test.tsx
Comment (hokei) [inline src/react-components/src/__tests__/components/Hero-test.tsx]: nit

**Review by hokei — COMMENTED** [review]
looks good, just a reminder to add a changelog

File: src/react-components/src/components/00-base/SectionName.tsx
Comment (hokei) [reply src/react-components/src/components/00-base/SectionName.tsx:4]: ah yes my memory is jogged now, sorry about that.

File: src/styles/03-organisms/headers/_hero.scss
Comment (helenvholmes) [reply src/styles/03-organisms/headers/_hero.scss]: Oops, old code—removing

File: src/styles/CHANGELOG.md
Comment (hokei) [inline src/styles/CHANGELOG.md]: Minor (? Since there's no release yet?) bump this and call out breaking changes.

**Comment by hokei** [comment]
I'm so sorry:  
Screen Shot 2020-04-10 at 9 35 17 AM
(The image is above the bar) 

Screen Shot 2020-04-10 at 9 35 25 AM
(the image is not)

Maybe the design has changed?

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx]: This particular image looks like there's a margin-left and margin-right to it.  Subtle, but since these are "reference components" maybe replace it with a clearer image?

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (hokei) [inline src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx]: Not requesting changes, just some Friday brain-teasers:  
ResearchNow has been renamed to Digital Research Books.  How might we keep something like this up-to-date?

**Comment by hokei** [comment]
at the same screen size: 

Screen Shot 2020-04-10 at 9 42 24 AM
Screen Shot 2020-04-10 at 9 42 29 AM

**Review by hokei — COMMENTED** [review]
super close!  The JS all looks good, just some visual issues found from pulling the code

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (EdwinGuzman) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx]: Try our best to not give any components specific names.

File: src/react-components/src/__tests__/components/Hero-test.tsx
Comment (EdwinGuzman) [inline src/react-components/src/__tests__/components/Hero-test.tsx]: You're not using `BodyText` in this file.

File: src/react-components/src/__tests__/components/Hero-test.tsx
Comment (EdwinGuzman) [inline src/react-components/src/__tests__/components/Hero-test.tsx]: These aren't being used.

File: src/react-components/src/__tests__/components/Hero-test.tsx
Comment (EdwinGuzman) [inline src/react-components/src/__tests__/components/Hero-test.tsx]: I would test that you can find an image in this DOM since the description says it has a background image. I'm guessing this is hard to test because a background image is css?

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (hokei) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx]: This is in stories, though ... I had been thinking one way Design System can move forward is to build out the "story" version of a component and pass it to the team that's supposed to consume it.  This is also the best solution I can think of to do regression testing (eg: if a component needs to change, how do we make sure it doesn't drastically affect existing users of the component?) But when the product changes, there's not really a good way to signal that back to Design System so that they can update the story ...

Comment (EdwinGuzman) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx]: Yeah, I noticed that after... I guess the person writing the stories should just keep the names up-to-date, and also remove any that are no longer used.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: These components aren't being used in this file.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: This variable isn't being used in this file.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx:66]: Should there be a fallback in case PRIMARY or SECONDARY aren't passed?

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx]: Not necessary but you can make line 86 have `= {}` as the default. That way you can remove these lines from 97-99.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (EdwinGuzman) [inline src/react-components/src/components/03-organisms/Hero/Hero.tsx:100]: I like how you did this responsive styling through the use of `data-*`.

**Comment by EdwinGuzman** [comment]
Update looks good. There are a couple of minor code changes you can do.

In the Hero tests, you caught 2 errors that can be thrown, but there are a total of 5 potential errors. It'd be good to test those cases.

File: src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.stories.tsx]: Name changes to products don't happen very often. When they do, we should create issues to track down any stories with an old product name and get them updated. I'll do that here.

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx:100]: Can't take credit for this, this is just how 4K did it originally :)

File: src/react-components/src/components/03-organisms/Hero/Hero.tsx
Comment (helenvholmes) [reply src/react-components/src/components/03-organisms/Hero/Hero.tsx:66]: I've instead made `heroType` required, so it'll now fail if a value from that enum isn't passed (it being optional was a holdover from when I was reworking this code).

**Comment by helenvholmes** [comment]
> I'm so sorry:
> Screen Shot 2020-04-10 at 9 35 17 AM
> (The image is above the bar)
> 
> Screen Shot 2020-04-10 at 9 35 25 AM
> (the image is not)
> 
> Maybe the design has changed?

Have a note to talk about it in [UX's Housekeeping meeting tomorrow](https://docs.google.com/document/d/16C5ZCAkTNt_n75FzDqvueaVjlVgDh9ye6C7O62z_yLo/edit#) to get the designs updated to what's built. The former designs just proved too challenging to build.

> at the same screen size:
> 
> Screen Shot 2020-04-10 at 9 42 24 AM
> Screen Shot 2020-04-10 at 9 42 29 AM

I believe this is working as intended (see [prod here](https://www.nypl.org/books-more/recommendations/staff-picks/adults)) but I think that the true determinate of this is Kathy, who can't do a VQA on this until this PR gets merged or I run a Zoom call with her. I think this particular VQA nit isn't worth blocking on.

File: src/react-components/src/__tests__/components/Hero-test.tsx
Comment (helenvholmes) [reply src/react-components/src/__tests__/components/Hero-test.tsx]: Yeah, I was struggling with it and left it to solve later... and obviously forgot about it. Figured it out and updated it

**Review by EdwinGuzman — APPROVED** [review]
The code looks good. I didn't look too closely at the designs and metronome because I don't feel too comfortable commenting on those updates.

**Comment by helenvholmes** [comment]
@EdwinGuzman long term I think it'd be best to get a multidev environment set up using Heroku or something so Kathy, Ellen, or Marcus can weigh in—I understand the reticence!

**Comment by helenvholmes** [comment]
Marking as `dependent` on @hokei's Filters PR.

File: src/react-components/CHANGELOG.md
Comment (hokei) [inline src/react-components/CHANGELOG.md:7]: Will also need a package.json update as well

**Review by hokei — APPROVED** [review]

**Comment by helenvholmes** [comment]
@hokei Can we look into shipping this next week if I do the integration, or would you prefer we continue prioritizing Template work?

**Comment by hokei** [comment]
Let's ship this first!

**Comment by helenvholmes** [comment]
Merged, although looks like Github being down caused some weirdness where it's not getting recognized. Closing this out.

**Pull request closed without merging**
